src/tests/multihost/alltests/test_ldap_password_policy.py

```diff
@@ -0,0 +1,258 @@
+            4. Should succeed
+        """
+        client = multihost.client[0]
+        tools = sssdTools(multihost.client[0])
+        tools.sssd_conf("nss", {'filter_groups': 'root',
+                                'filter_users': 'root',
+                                'debug_level': '9'}, action='update')
+        ldap_params = {'use_fully_qualified_names': False}
+        tools.sssd_conf(f'domain/{ds_instance_name}', ldap_params)
+        client_hostname = multihost.client[0].sys_hostname
+        # setup 389-ds w/server-side pw controls
+        cn_config = 'cn=config'
+        for element, value in [('passwordExp', [b'on']),
+                               ('passwordMaxAge', [b'86400']),
+                               ('passwordWarning', [b'86400'])]:
+            ldap_modify_ds(multihost, ldap.MOD_REPLACE, cn_config, element, value)
+        # set the password
+        user_dn = 'uid=ppuser1,ou=People,dc=example,dc=test'
+        ldap_modify_ds(multihost, ldap.MOD_REPLACE, user_dn, 'userPassword', [b'Secret123'])
+        client.run_command('> /var/log/secure')
+        tools.clear_sssd_cache()
+        auth_success(client_hostname, 'ppuser1', 'Secret123')
+        time.sleep(3)
+        file_scure = '/var/log/secure'
+        file_ssd = f'/var/log/sssd/sssd_{ds_instance_name}.log'
+        find_logs(multihost, file_scure, "Your password will expire in ")
+        find_logs(multihost, file_ssd, "Server returned control [1.3.6.1.4.1.42.2.27.8.5.1]")
+        find_logs(multihost, file_ssd, "Password will expire in [86")
+        with pytest.raises(AssertionError):
+            find_logs(multihost, file_ssd,
+                      "Server does not support the requested "
+                      "control [1.3.6.1.4.1.42.2.27.8.5.1]")
+
+    @staticmethod
+    def test_maxage(multihost):
+        """
+        :title: "Set passwordMaxAge to 1"
+        :id: fc68293e-9676-11ed-9808-845cf3eff344
+        :steps:
+            1. setup 389-ds w/server-side passwordMaxAge to 1
+            2. Check user was to change password
+        :expectedresults:
+            1. Should succeed
+            2. Should succeed
+        """
+        client = multihost.client[0]
+        tools = sssdTools(multihost.client[0])
+        client_hostname = multihost.client[0].sys_hostname
+        cn_config = 'cn=config'
+        ldap_modify_ds(multihost, ldap.MOD_REPLACE, cn_config, 'passwordMaxAge', [b'1'])
+        user_dn = 'uid=ppuser1,ou=People,dc=example,dc=test'
+        ldap_modify_ds(multihost, ldap.MOD_REPLACE, user_dn, 'userPassword', [b'Secret123'])
+        client.run_command('> /var/log/secure')
+        tools.clear_sssd_cache()
+        file_location = '/script/change_user_password_while_expired.sh'
+        client.transport.put_file(os.path.dirname(os.path.abspath(__file__))
+                                  + file_location, '/tmp/change_user_password_while_expired.sh')
+        client.run_command('sh /tmp/change_user_password_while_expired.sh')
```

Comment:
I wonder if change_user_password from utils.py would not work.

Reply:
It will not wok because of : 

https://github.com/SSSD/sssd/blob/master/src/tests/multihost/sssd/testlib/common/utils.py#L675

user will never get to this point which is change_user_password is expecting . That is the reason i am using this script

Reply:
THis is what you will get.

(Pdb) tools.change_user_password('ppuser1', 'Secret123', 'Secret123', 'NewPass_123', 'NewPass_123')
spawn ssh -o NumberOfPasswordPrompts=1 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -l ppuser1 localhost
expect "*assword: "
"end "Secret123
expect "$ "
"end "passwd
expect "Current password: "
"end "Secret123
expect "New password: "
"end "NewPass_123
expect "Retype new password:"
"end "NewPass_123
expect {
	timeout { set result_code 0 }
	"passwd: all authentication tokens updated successfully" { set result_code 3 }
	"passwd: Authentication token is no longer valid; new one required" { set result_code 4 }
	"Sorry, passwords do not match." { set result_code 5}
	"Password change failed. Server message: Old password not accepted." { set result_code 6 }
	eof {}
	"Permission denied " { set result_code 10 }
}
exit $result_code

2023-01-25 14:09:23,369 - sssd.testlib.common.qe_class.QeHost.clientrel9A.ParamikoTransport - INFO - WRITE /tmp/qe_pytest_expect_fileKNFQLV3UMD
remote side expect script filename: /tmp/qe_pytest_expect_fileKNFQLV3UMD
2023-01-25 14:09:23,376 - sssd.testlib.common.qe_class.QeHost.clientrel9A.ParamikoTransport - INFO - RUN ['expect', '-f', '/tmp/qe_pytest_expect_fileKNFQLV3UMD']
2023-01-25 14:09:23,376 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - RUN ['expect', '-f', '/tmp/qe_pytest_expect_fileKNFQLV3UMD']
2023-01-25 14:09:23,461 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - -bash: line 1: cd: /root/multihost_tests: No such file or directory
2023-01-25 14:09:23,461 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - -bash: line 2: /root/multihost_tests/env.sh: No such file or directory
2023-01-25 14:09:23,465 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - spawn ssh -o NumberOfPasswordPrompts=1 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -l ppuser1 localhost
2023-01-25 14:09:23,502 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - Warning: Permanently added 'localhost' (ED25519) to the list of known hosts.
(ppuser1@localhost) Password: .testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - 
(ppuser1@localhost) Password expired. Change your password now.lientrel9A.cmd121 - DEBUG - 
2023-01-25 14:09:33,601 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - Current Password: 
2023-01-25 14:09:33,650 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - Password change failed. Server message: Old password not accepted.
2023-01-25 14:09:33,650 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - 
ppuser1@localhost's password: .testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - 
2023-01-25 14:09:43,779 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - Password expired. Change your password now.
2023-01-25 14:09:43,914 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - Password expired. Change your password now.
2023-01-25 14:09:43,914 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - This system is not registered to Red Hat Insights. See https://cloud.redhat.com/
2023-01-25 14:09:43,914 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - To register this system, run: insights-client --register
2023-01-25 14:09:43,914 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - 
2023-01-25 14:09:43,914 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - Last failed login: Wed Jan 25 14:09:33 IST 2023 from 127.0.0.1 on ssh:notty
2023-01-25 14:09:43,914 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - There was 1 failed login attempt since the last successful login.
2023-01-25 14:09:43,915 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - Last login: Wed Jan 25 14:04:30 2023 from 172.16.162.1
2023-01-25 14:09:43,915 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - WARNING: Your password has expired.
2023-01-25 14:09:43,915 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - You must change your password now and login again!
2023-01-25 14:09:43,925 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - Changing password for user ppuser1.
2023-01-25 14:09:53,933 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - Current Password: 
2023-01-25 14:09:53,976 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - Password change failed. Server message: Old password not accepted.
2023-01-25 14:09:55,860 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - passwd: Authentication token manipulation error
2023-01-25 14:09:55,864 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - Connection to localhost closed.
2023-01-25 14:09:55,865 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - send: spawn id exp3 not open
2023-01-25 14:09:55,865 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG -     while executing
2023-01-25 14:09:55,865 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - "send "NewPass_123
2023-01-25 14:09:55,865 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - ""
2023-01-25 14:09:55,865 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG -     (file "/tmp/qe_pytest_expect_fileKNFQLV3UMD" line 15)
2023-01-25 14:09:55,866 - sssd.testlib.common.qe_class.QeHost.clientrel9A.cmd121 - DEBUG - Exit code: 1
----expect output start----
spawn ssh -o NumberOfPasswordPrompts=1 -o StrictHostKeyChecking=no -o UserKnownHostsFile=/dev/null -l ppuser1 localhost
Warning: Permanently added 'localhost' (ED25519) to the list of known hosts.
(ppuser1@localhost) Password: 
(ppuser1@localhost) Password expired. Change your password now.
Current Password: 
Password change failed. Server message: Old password not accepted.

ppuser1@localhost's password: 
Password expired. Change your password now.
Password expired. Change your password now.
This system is not registered to Red Hat Insights. See https://cloud.redhat.com/
To register this system, run: insights-client --register

Last failed login: Wed Jan 25 14:09:33 IST 2023 from 127.0.0.1 on ssh:notty
There was 1 failed login attempt since the last successful login.
Last login: Wed Jan 25 14:04:30 2023 from 172.16.162.1
WARNING: Your password has expired.
You must change your password now and login again!
Changing password for user ppuser1.
Current Password: 
Password change failed. Server message: Old password not accepted.
passwd: Authentication token manipulation error
Connection to localhost closed.

-bash: line 1: cd: /root/multihost_tests: No such file or directory
-bash: line 2: /root/multihost_tests/env.sh: No such file or directory
send: spawn id exp3 not open
    while executing
"send "NewPass_123
""
    (file "/tmp/qe_pytest_expect_fileKNFQLV3UMD" line 15)

----expect output end----
1